tests/programs/custom/clic/clic.c

```diff
@@ -2161,49 +2161,54 @@ uint32_t rw_mscratchcsw(uint32_t index, uint8_t report_name) {
     return 0;
   }
 
-  __asm__ volatile (R"(
-    csrrs %[rd1], mstatus, zero
-  )":[rd1] "=r"(mstatus_rval.raw)
-    ::);
-
   if (has_umode_configured()) {
-    mstatus_rval.fields.mpp = 0x0;
+    __asm__ volatile (R"(
+      csrrs %[rd1], mstatus, zero
+    )":[rd1] "=r"(mstatus_rval.raw)
+      ::);
+
+    if (has_umode_configured()) {
+      mstatus_rval.fields.mpp = 0x0;
+
+      // Set mpp to zero and attempt swap
+      __asm__ volatile (R"(
+        csrrw %[rd1], mstatus, %[rd1]
+        add   %[rd2], sp, zero
+        csrrw %[rd3], 0x348, sp
+        csrrs %[rd4], mscratch, zero
+        csrrw sp, 0x348, %[rd3]
+        add   %[rd3], sp, zero
+        csrrw zero, mscratch, zero
+      )":[rd1] "+r"(mstatus_rval.raw),
+         [rd2] "=r"(reg_backup_1),
+         [rd3] "+r"(reg_backup_2),
+         [rd4] "=r"(mscratch)
+        ::);
```

Comment:
This is difficult for me to review, because I don't know clic very well, and because... assembly is difficult.
So I will simply trust it.
The code gets put to the test, so there is assurance in the self-checking.